### ISSUE TITLE
[selectors] Pseudo-class to indicate when a slot has content

### ISSUE THREAD

**Issue by plinss** [w3c/csswg-drafts#6867]
I'm developing several web components and I'm finding a need to tell when a slot has content or not. (e.g. to apply margins or not so empty slots don't impact layout, but layout well when they do, etc.)

The :empty pseudo-class applies when the slot doesn't have direct content, even when it has slotted content, so doesn't work. Using slot:has(::slotted(*)) wouldn't detect text-only content.

Once possibility would be to redefine :empty to detect slotted content, but it may be useful to keep the current behavior to differentiate empty slots from slots with default content (and the change may break exiting content). So a new pseudo-class that only detects slotted content may be useful.

**Comment by WickyNilliams** [comment]
I have also needed this on several occasions. At the moment, you need to (at the minimum) inspect `assignedNodes`/`assignedElements` to determine this, and then add css classes to elements to modify styling. This is far from ideal for such a common use case. 

A typical use case for me is to hide (`display:none`) a `<slot>` if it is has no assigned content. The reason being that i often set a slot's `display` to something other than `contents` e.g. making a slot a flex container so that i can use `gap` to space slotted children. 

So being able to target a slot with assigned content is the most pressing use-case. Though I think it's also worth considering whether there should be separate pseudo-class for each of these states:

* assigned content
* default content
* empty (i.e. no default or assigned content)

**Comment by MaxArt2501** [comment]
I'd like to point out that a `:has-assigned-content` pseudo-class (or better name) would be nice to have since:
* developers would like to avoid using JavaScript for what's basically a declarative effect;
* working around with `:host(:not(:empty)) slot` (for unnamed slots) or `:host(:has([slot='foo'])) slot[name='foo']`:
  * is verbose;
  * has content limitations (e.g. for components with multiple slots);
  * doesn't work for manual node assignment.

I'm quite keen myself to set a `display` value different than `contents` for slots, but this would also allow to style sibling or parent nodes (via `:has()`).

**Comment by castastrophe** [comment]
Perhaps inline with `:has`, `:has-assigned` (content could be a misnomer because it might contain text or nodes). This would enable styles such as:

```
slot[name='foo']:not(:has-assigned) {
   display: none;
}
```

A good partner to this spec would be #6620

**Comment by castastrophe** [comment]
This seems like a related request in the WICG space: https://github.com/WICG/webcomponents/issues/936

**Comment by AutoSponge** [comment]
in the example the `::slotted(*)`, the `*` is intended to be a selector but it may have more benefit to developers if the slot name was also queryable in a similar manner. I'm thinking of cases where a component uses multiple sibling slots.

**Comment by tabatkins** [comment]
Can you elaborate on what you're thinking of? The `::slotted()` pseudo-element lives *on* a `slot` element, so you can already distinguish between multiple slots.  The selector argument is matched against the elements put *into* the slot.

----

As I said in the minutes above, I think `:has-slotted` is a reasonably straightforward name - not too long, but clearly targeted at its use-case.

**Comment by Westbrook** [comment]
@tabatkins where are you seeing the `:has-slotted` selector being available? Feels like the optimal usability of it would benefit from it being available all the way up to `:host:has-slotted([slot=name])`.

Would be we able to do something here along the lines of [`.assignedElements({flatten: true})`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/assignedElements#flatten) to ensure this selector is pointing to "content" and not just "possible content" (e.g. `<slot>` elements)?

Is there any benefit in being forward looking to the discussion in https://github.com/w3c/csswg-drafts/issues/7922 and make it a combinator? Maybe that doesn't make sense in this case...

**Comment by castastrophe** [comment]
An interesting edge case to consider will be slots with default content as well. Does default content could as `has-slotted`? i.e.,

```
<slot>
    <p>Lorem ipsum dolor...</p>
</slot>
```

**Comment by MaxArt2501** [comment]
> An interesting edge case to consider will be slots with default content as well. Does default content count as `has-slotted`? i.e.,
> 
> ```
> <slot>
>     <p>Lorem ipsum dolor...</p>
> </slot>
> ```

I'd say no, because that `<p>` isn't technically slotted. One can combine `:has-slotted` with `:empty`, if necessary.

But perhaps there may be more use cases.

**Comment by Westbrook** [comment]
Great nuance in:
> ```
> <slot>
>    <p>Lorem ipsum dolor...</p>
> </slot>
> ```
That `p` comes back in `.assignedElements({flatten: true})` but not `.assignedElements()` and is _NOT_ available in `::slotted(p)` but is in `:host p`, which vaguely implies that we'd need two selectors or rather one with some settings (!? 😳, not sure if there are examples of that beyond [attribute selectors](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors#attr_operator_value_i)) to ensure we have full control over the content we're shipping here.

I'm not quite sure how you could combine `:has-slotted` and `:empty` to get the sort of outcomes desired here, specifically in that `:empty` is an element state selector and `:has-slotted` is a child state selector, they don't apply to the same elements in my imagination. Could you share more on that? I could maybe see `:not(:has-slotted(...))` getting you somewhere (assuming the `p` wasn't queried by this selector), but then how would you capture that the `p` is there, unless we can also have `:host:has` (point of order: not in the function reference, but directly on the element) that pointed into the shadow root...interesting.

**Comment by MaxArt2501** [comment]
Now that I think of it, it's not just `:empty` that could come in handy, but also `:has()`. If I want to hide a slot that has no slotted node *and* no default content, I could achieve it with

```css
slot:not(:has(*)):not(:has-slotted(*)) {
  display: none;
}
```

(`:empty` could be useful when there's just default *text* as content, like `<slot>Hello</slot>`... alas, it's brittle enough that it'd fail with just some whitespace.) `:has` and `:has-slotted` should cover all the basic uses. For example, if we have something like this, where that content can be wither passed as an attribute (for simple text) or as slotted content (for HTML content):

```html
<slot name="title">${this.cardTitle}</slot>
```

I wasn't aware that default slot content is returned by `.assignedElements({flatten: true})`, as considering it "assigned" to any slot seems counter-intuitive to me. All in all, I'm not sure that trying to mimic the behavior of a JS method would make sense here.

**Comment by tabatkins** [comment]
> where are you seeing the :has-slotted selector being available? Feels like the optimal usability of it would benefit from it being available all the way up to :host:has-slotted([slot=name]).

It's on the `slot`, indicating that that particular `slot` element has slotted content in it. If you want to style the host based on whether any of its slots have slotted content, you can use `:has(slot:has-slotted)` to check. ^_^

> An interesting edge case to consider will be slots with default content as well. Does default content count as has-slotted?

I'd say no. It's not slotted content, and in particular, it's not useful to *detect* slotted content in this case if this is true. That is, if `:has-slotted` matches a slot with default contents, then what would one do with the selector? It becomes *guaranteed* to match, and at that point you can just put a class on the `slot` and use that instead. On the other hand, having it *not* match when the slot is using its default content lets you make a meaningful distinction that you can only know at runtime, which is useful.

**Comment by justinfagnani** [comment]
We have a number of SSR users who really need this selection.

There a few reasons this is critical in some SSR implementations:
 1. They may not actually use a DOM on the server - they're just emitting strings to the HTTP response. So there may be no way to determine if an element slots into a parent's slot.
 2. The current workaround of listening for `slotchange` and setting a class uses imperative DOM code that doesn't work on the server, and doesn't work on initial hydration without JS.
 
 It seems like the idea of a new selector is fairly well received... can this idea be added to an upcoming csswg meeting?

**Comment by claviska** [comment]
In the case of [Shoelace](https://shoelace.style/), we're using [imperative DOM code](https://github.com/shoelace-style/shoelace/blob/854db13bd7fbfad65f01de0838fe1a6b2dbf0a4d/src/internal/slot.ts#L3-L62) to do slot detection which, as @justinfagnani mentioned, doesn't work on the server. We can _probably_ work around some of these checks by changing structure/display so empty slots collapse, but this is very uncomfortable and won't solve all of our use cases.

As it stands, without such a selector, we can't get SSR fully functional. This has, unfortunately, been a pretty big hurdle for adopters who want to use our custom elements with metaframeworks + SSR.

**Comment by michaelwarren1106** [comment]
I'm using the same imperative code mechanism as Shoelace above to detect slots having content. A css selector would be amazing.

**Comment by Westbrook** [comment]
100%. Even though we’re not yet addressing SSR’d content at Adobe, selectors of this shape point to 5-15% rendering benefits for elements that leverage this type of functionality as powered by JS. Looking forwards to more options here!

**Comment by andyearnshaw** [comment]
I'd also like to see the function-style `:has-slotted(*)`, too. I'm building out some components where the default styling for the host is `display: flex`. The `<slot>` elements are direct children of the shadow root and become part of the `flex` layout. However, this causes a complication where the consumer is forced to wrap text nodes in an element in order to get the correct layout on them. Since `:has-slotted(*)` would match a slot with anything but text nodes slotted, I could do something like this:

```css
.my-slot:not(:has-slotted(*)) {
  display: block;
}
```

I feel this gives a little more flexibilty and power to make custom elements a little more ergonomic. There are potential other use cases, such as internal shadow tree styling based on an invalid slotted form field or an `<a>` with a missing `href` attribute, for example.

**Comment by Westbrook** [comment]
@tabatkins @fantasai how can the Web Components Community Group support yall in fulfilling the resolution made [above](https://github.com/w3c/csswg-drafts/issues/6867#issuecomment-1332344714) to bring a draft proposal to this situation?

This is a great amount of excitement for the ability to detect slot content in the CSS context within the community and look forward to helping to get this added to the web platform!

**Comment by sorvell** [comment]
Seems like this is somewhat possible with style queries given cooperation from the outer scope.

https://lit.dev/playground/#gist=e21987101802ecf8c3381a0ac11dca1d

**Comment by trusktr** [comment]
Will `:has-slotted` match when the slot forwards to another deeper slot? We will need a way to distinguish *finally* slotted content vs content that is being forwarded. We might want to want to style a slot or its slotted content

- if the slot participates in slotting even if that content is forwarded to a deeper slot, f.e. with `:has-slotted`
- only if the slot is the *final* destination of any content, f.e. with `:has-slotted-final` or something

**Comment by jpzwarte** [comment]
I just want to do `:has-slotted(:focus-visible) { outline: .... };` in my web component that has an `<input type="text">` in the light DOM 🙏

**Comment by LeaVerou** [comment]
Circling back to this, as this is a big pain point for several WC use cases.
It appears this may require more discussion:

- The [resolution](https://github.com/w3c/csswg-drafts/issues/6867#issuecomment-1332344714) mentions Selectors 5, but surely at this point css-scoping is the right place?
- OP was requesting a non-functional pseudo-class, but use cases emerged for a functional form too. Note that as mentioned in the original request, the non-functional form is not simply an alias to `:has-slotted(*)`, as the latter **does not detect text nodes**.
- There have been questions about what syntax should be allowed for the argument, and whether this accounts for flattening

So I’m Agenda+ing so we can make some progress next week.

**Comment by justinfagnani** [comment]
> OP was requesting a non-functional pseudo-class, but use cases emerged for a functional form too.

This thread might have requested a non-funciton form, but the issue I opened before (https://github.com/WICG/webcomponents/issues/936) definitely requested a function form so that `:has-slotted()` matches whenever the corresponding `::slotted()` would have matched something. I think the functional form is important for consistency and some of the use-cases I've come across.

**Comment by LeaVerou** [comment]
@justinfagnani Agreed, but I believe the resolution was for the non-functional form, which has its own, distinct, use cases. I agree the functional form is worth pursuing, but needs more design discussion, and shouldn't hold back the simpler, non-functional form.

**Comment by keithamus** [comment]
I don't have particularly strong opinions (and I'm not an implementor, though I did have a [go at implementing `:has-slotted()` in WebKit](https://github.com/WebKit/WebKit/compare/main...keithamus:wip-has-slotted) and intend to do as much as I can to get this specified and implemented across the major engines), but it seems as though if `:has-slotted` is an effective alias for `:has-slotted(*)`, then the leap to defining `:has-slotted()` feels like a reasonably tractable step to tackle both rather than one at a time.

At any rate I'm happy to see this labelled with https://github.com/w3c/csswg-drafts/labels/Agenda%2B%20TPAC and look forward to the discussion and to learn the next useful steps to take.

**Comment by Westbrook** [comment]
Discussing at TPAC sounds excellent! Hopefully, there's not too much double booking and I can join remotely.

To ensure all the data points are collected, I proposed these WPTs in relation to this feature: https://github.com/web-platform-tests/wpt/pull/47587/files

Not having any other input, we (the Web Component Community Group live coded this as a group) specifically marked non-functional `:has-slotted` as _invalid_, in parallel to the invalidity of non-functional `::slotted`. While allowing for non-functional `:has-slotted` could be useful in some cases, as both a custom element developer and the chair of the community group, there's not nearly as much (if any) call for that as there is for the functional version.

At the pace of this, beggars certainly can't be choosers, but as I said, wanted to make sure as many data points were collected as possible.

**Comment by justinfagnani** [comment]
@keithamus 

> if `:has-slotted` is an effective alias for `:has-slotted(*)`

It isn't though. @LeaVerou's point here is very important:

> * the non-functional form is not simply an alias to `:has-slotted(*)`, as the latter **does not detect text nodes**.

**Comment by LeaVerou** [comment]
Non-functional `::slotted` would make no sense: what is it targeting? Non-functional `:has-slotted` not only makes sense, it has its own, distinct use cases. Moreover, both in my and @claviska’s experience building custom elements, the vast majority of use cases we've come across simply required knowing if any nodes were assigned to the slot at all, whether text nodes or anything else. These were also the use cases @plinss opened this issue for.

----

@claviska also reminded me of a potential footgun here: does a single whitespace text node match `:has-slotted`? It does become slotted, but that’s a footgun in the current design, see https://x.com/LeaVerou/status/1785904086929346957
Do we reproduce this footgun or break [consistency](https://www.w3.org/TR/design-principles/#consistency)?

**Comment by justinfagnani** [comment]
> both in my and @claviska’s experience building custom elements, the vast majority of use cases we've come across simply required knowing if any nodes were assigned to the slot at all, whether text nodes or anything else.

How many of these were for named slots? You can only slot elements into named slots, so `:has-slotted` and `:has-slotted(*)` would be equivalent in those cases.



> does a single whitespace text node match `:has-slotted`? It does become slotted, but that’s a footgun in the current design, see https://x.com/LeaVerou/status/1785904086929346957
> Do we reproduce this footgun or break [consistency](https://www.w3.org/TR/design-principles/#consistency)?

I think you have to reproduce this behavior. An important aspect to consistency is fallback content. A whitespace text node will prevent fallback content from showing. I don't think we want `:has-slotted` to disagree with fallback.

**Comment by keithamus** [comment]
> @keithamus
> 
> > if `:has-slotted` is an effective alias for `:has-slotted(*)`
> 
> It isn't though. @LeaVerou's point here is very important:
> 
> > * the non-functional form is not simply an alias to `:has-slotted(*)`, as the latter **does not detect text nodes**.

Apologies I misread. In that both make sense to have, but again (hubris incoming) I think they can both be specified and implemented reasonably straightfowardly in unison, it's certainly how I'd prefer to tackle it.

**Comment by sorvell** [comment]
> does a single whitespace text node match :has-slotted?

Agree that this is a nasty foot gun that needs to be addressed. It's not exclusively a CSS issue since it prevents fallback content from displaying. Ideally we'd solve the core issue and _not_ just add a CSS bandaid, but I propose the following in the interests of making concrete progress:

1. keep `has-slotted(...)`: and you can combine with `:not(:has-slotted(...))` to detect elements not selected
2. add `has-slotted-empty`: this matches the same as :empty is [specified](https://drafts.csswg.org/selectors-4/#the-empty-pseudo) and not as it's [implemented](https://developer.mozilla.org/en-US/docs/Web/CSS/:empty) (assuming it will be possible to actually match spec?). **This means it *would* match if whitespace-only nodes are assigned.**

If we don't get this right, the following would show a red box and that seems, to me, a bit absurd.

```html
<x-foo>
  <template shadowrootmode="open">
    <style>
      slot:has-slotted { 
       display: block; 
       height: 20px; 
       background: red;
      }
    </style>
    <slot></slot>
  </template>
</x-foo>
```

**Comment by LeaVerou** [comment]
Unfortunate as it may be, I am tending to agree that consistency is more important, and we need to reproduce the JS behavior, while providing some kind of mechanism to at least make these cases detectable. 😢 What that mechanism would be needs more design discussion `:has-slotted-empty` reads very weird to me, and I wouldn't base things on entirely unimplemented precedent).

@keithamus looking at the PR, it seems there are many more design decisions to be made with the functional version, which would make the non-functional version at least easier to specify, if not also easier to implement. That doesn't mean we won't work on the functional version — if anything getting the non-functional version in ASAP paves the way for working on the functional version, it doesn’t compete against it.

**Comment by tabatkins** [comment]
@trusktr 
> Will :has-slotted match when the slot forwards to another deeper slot? We will need a way to distinguish slots with finally slotted content vs slots with content that is being forwarded to a deep slot.

Could you provide an example of that? I didn't think about the distinction, so I'm not sure.

--------

@justinfagnani 
> I think you have to reproduce this behavior. An important aspect to consistency is fallback content. A whitespace text node will prevent fallback content from showing. I don't think we want :has-slotted to disagree with fallback.

Agree, tho I agree with Lea that it sucks. But yeah, I believe the major use-case of `:has-slotted` is giving it different styling based on whether the fallback is showing or not, so it needs to be consistent. Ideally we should be able to specify that a slot needs *elements* (or at least significant text) so it'll be fixed in all the relevant APIs at once.

In the absence of that, I think @sorvell's idea of `:has-slotted-empty` has merit; it's a patch, but it's simple.

--------

re: functional form, agreed with Lea. The resolution was for the non-functional form. We can pursue the functional form *as well*, but it can be done separately; CSS is fine with handling a pseudo with both forms.

**Comment by sorvell** [comment]
@LeaVerou 

> Unfortunate as it may be, I am tending to agree that consistency is more important, and we need to reproduce the JS behavior

Yeah, in the ideal case I completely agree, but I worry that the fact that the behavior spans CSS and JS makes our typically glacial pace more cosmological here. It also seems like we're down the path of inconsistency already with `:has-slotted` instead of `:has(:slotted)`, and this is not a reason to continue in this direction. I'm also really unsure on that status of `:empty`. Matching what's spec'd there seemed appealing, but that's definitely less the case if the spec is likely to adapt to what's implemented here.

**Comment by tabatkins** [comment]
> It also seems like we're down the path of inconsistency already with :has-slotted instead of :has(:slotted)

That's not at all inconsistent. You can't do `slot > .slotted-element` to select things slotted in, so `slot:has(.slotted-element)` won't work either. You need a new concept to be able to match that stuff.

**Comment by LeaVerou** [comment]
@sorvell That is not an inconsistency, `:has(:slotted)` has an existing meaning that doesn’t fit what we want to do here: it would match **descendants** that match a (nonexistent) `:slotted` pseudo-class. 

I doubt `:empty` will ever be fixed at this point as that would have web compat implications — we'd probably need to define a new one at this point. Not sure if anyone is already working on this.

**Comment by sorvell** [comment]
> That's not at all inconsistent.

Yup, my mistake. It is indeed getting hard to parse these functions =)

**Comment by LeaVerou** [comment]
Talking to @emilio yesterday, he confirmed my suspicion that `:has-slotted` is indeed way easier to implement than `:has-slotted()`. So I would focus on getting that specced and shipped ASAP and getting a resolution to work on `:has-slotted()` in a separate issue.

**Comment by LeaVerou** [comment]
Update: I submitted PR #10959 yesterday, @tabatkins reviewed it today, and `:has-slotted` has now landed: https://drafts.csswg.org/css-scoping/#the-has-slotted-pseudo

I’ve also had chats with @emilio and @rniwa about implementation difficulty and they both said it’s trivial to do. Not sure who the Blink implementor is (@chrishtr @mfreed7 do you know?), but I imagine the answer would be similar. So, fingers crossed this could ship relatively soon, as it’s a nice quick win 🤞🏼

**Comment by mfreed7** [comment]
> Update: I submitted PR #10959 yesterday, @tabatkins reviewed it today, and `:has-slotted` has now landed: https://drafts.csswg.org/css-scoping/#the-has-slotted-pseudo 🎉
> 
> I’ve also had chats with @emilio and @rniwa about implementation difficulty and they both said it’s trivial to do. Not sure who the Blink implementor is (@chrishtr @mfreed7 do you know?), but I imagine the answer would be similar. So, fingers crossed this could ship in browsers relatively soon, as it’s a nice quick win 🤞🏼

My team could take a look - sounds rather straightforward to implement. A few questions/requests:
 1. Are there WPTs?
 2. Would you mind [filing an implementation bug](https://crbug.com/new) for us?

**Comment by keithamus** [comment]
@mfreed7 happy to implement in Chrome. We have some WPTs, I'll file a bug.

**Comment by LeaVerou** [comment]
Thank you @mfreed7!

> 1. Are there WPTs?

Yes, though many are mixed with `:has-slotted()`. Here are the ones that include a bare `:has-slotted`:
- https://github.com/web-platform-tests/wpt/blob/master/css/css-scoping/has-slotted-007.tentative.html
- https://github.com/web-platform-tests/wpt/blob/master/css/css-scoping/has-slotted-002.tentative.html
- https://github.com/web-platform-tests/wpt/blob/master/css/css-scoping/has-slotted-flattened-002.tentative.html
- https://github.com/web-platform-tests/wpt/blob/master/css/css-scoping/has-slotted-flattened-007.tentative.html

Ideally the functional syntax should be separated into separate tests _even if_ we resolve to add it. They are separate features and need to be tested separately.

> 2. Would you mind [filing an implementation bug](https://crbug.com/new) for us?

Looks like someone beat me to it 15 hours ago! https://issues.chromium.org/issues/369883705

**Comment by mfreed7** [comment]
Thank you both! Looks like there's a lot of enthusiasm for the new pseudo class. Sounds like we're off and running.

> @mfreed7 happy to implement in Chrome. We have some WPTs, I'll file a bug.

In particular, thanks for this one! I can provide reviews, if needed.

**Comment by keithamus** [comment]
I think https://github.com/w3c/csswg-drafts/issues/6466 is the related issue - it talks of specificity but it looks representative of the cascade order issue.

**Comment by sorvell** [comment]
[Tested](https://lit.dev/playground/#gist=15f7e26306e5a6a65918def2c838eaaf) `:has-slotted` in Chrome Canary with the "experimental web platform features" setting on. 

The behavior is good for simple cases but not useful for nested composition. Perhaps this can be solved via https://github.com/w3c/csswg-drafts/issues/10771.

| Content | Expected | Actual | assignedNodes() | assignedNodes({flatten: true}) | :has-slotted works | ::slotted(*) |
|---------|:--------:|:------:|:---------------:|:------------------------------:|:------------------:|:------------:|
| `<glow-stuff></glow-stuff>` | no&nbsp;glow | no&nbsp;glow | `[]` | `[]` | yes | no&nbsp;match |
| `<glow-stuff><input></glow-stuff>` | glow | glow | `[<input>]` | `[<input>]` | yes | match |
| `<template shadowRootMode="open"><glow-stuff><slot></slot></glow-stuff></template>` | no&nbsp;glow | glow | `[<slot>]` | `[]` | no | no&nbsp;match |
| `<template shadowRootMode="open"><glow-stuff><slot></slot></glow-stuff></template><input>` | glow | glow | `[<slot>]` | `[<input>]` | by&nbsp;luck | match |
| `<template shadowRootMode="open"><glow-stuff><slot><input></slot></glow-stuff></template>` | glow | glow | `[<slot>]` | `[<input>]` | by&nbsp;luck | no&nbsp;match |

**Comment by keithamus** [comment]
I'm going to https://github.com/w3c/csswg-drafts/labels/Agenda%2B as I'd like to capture a resolution that the current behaviour of _not_ matching the _flat_ tree is okay, and perhaps to resolve to investigating a way to match the flat tree (though perhaps this is `:has-slotted()` (https://github.com/w3c/csswg-drafts/issues/7922) or a `/slotted/` combinator (https://github.com/w3c/csswg-drafts/issues/7922).

If web developers could please provide signal that the current behaviour as prototyped in Firefox and Chrome (that `:has-slotted` matches the non-flattened tree) is acceptable, that would help.

**Comment by Westbrook** [comment]
I really like having _anything_ available in this area (is there any way that I can further explain the **high level** of important the web component developing community places on this API?). I am super excited about the implementation currently available in Chrome/Firefox (so many things we could never do have already been opened up to us with this API). However, not being able to select against the flattened tree will be a hard blocker on this API; maybe not today, maybe not tomorrow, but it will come up (much like questions around white space being "slotted") that would be better to be answered today, if only directionally.

Personally, I can certainly be appeased with the selector matching `.assignedNodes({flatten: true})` _only_. For my experience, the ability to build complex, deeply slotted applications is and will continue to be greatly constrained by not supporting this. This also aligns it more directly with the behavior of `::slotted()`.  However, `::slotted()` does present a questionable result in this example:

```html
<div><template shadowrootmode="open">
  <div>
    <template shadowrootmode="open">
      <style>
        ::slotted(input) {
          color: red;
        }
      </style>
      <slot></slot>
    </template>
    <slot><input value="not styled" /></slot>
  </div>
</template></div>
```

_Practically, we do need to support the spread._

### Questions:

**Does that mean multiple selectors?** `:has-slotted` and `:has-slotted-flatten` could be a path, though it's interesting to need to go between the two. 

**Does `flatten` belong in the functional version?** Always? Or as an option? `:has-slotted(.selector !flatten)` _could_ be a path, but how this would apply to slotted text content is unclear. Also, does this mean each selector in the method get this option? `:has-slotted(.selector !flatten, .other-selector)`, see `.other-selector` not feature flatten. Or would this need to be general to a single `:has-slotted()` usage?

**Functionally, will component authors be doomed/required to "know" the slotted content depth?**

```html
<div>
  <template shadowrootmode="open">
    <div>
      <template shadowrootmode="open">
        <div>
          <template shadowrootmode="open">
            <style>
               p {
                 color: red;
               }
               slot:has-slotted(slot:has-slotted(slot:has-slotted(p))) + p {
                 color: green;
              }
            </style>
            <slot></slot>
            <p>Styled content</p>
          </template>
          <slot></slot>
        </div>
      </template>
    </div>
  </template>
  <p>Slotted content.</p>
</div>
```

**Does a combinator approach mean that `/slotted/` and `/slotted-flatten/` could be made possible together, leveraged with better x-browser support for [`:host:has(...)`](https://wpt.fyi/results/css/css-scoping?label=master&label=experimental&aligned&q=host%20has%20internal), and clarify both this and other situations herein?** Can the combinator select on text? 

**Could any of these practically address selecting when content is delivered via the hole in `::slotted()` outlined above?**

---

Many thanks to all who are investing time into making this a high-quality and powerful additions to the web platform! 🙇🏼‍♂️

**Comment by sorvell** [comment]
@keithamus 

> I'm going to [Agenda+](https://github.com/w3c/csswg-drafts/labels/Agenda%2B) as I'd like to capture a resolution that the current behaviour of not matching the flat tree is okay

Based on the 3rd row of the [table here](https://github.com/w3c/csswg-drafts/issues/6867#issuecomment-2395119379), it's my opinion that:

**The bare form of `:has-slotted` should match against the same tree (text node inclusive) as `::slotted(...)`.**  

Note, it's slightly confusing to call this the flattened tree because `assignedNodes({flatten: true})` contains fallback content but `::slotted` does not match it.

The basic use case for this: 
1. I have a button that needs different styling based on the presence of an icon. I will use `:has-slotted` to achieve this.
2. If I want to allow a user of my composite element to take advantage of this feature, I'd do: `<my-button><slot name="icon"></slot></my-button>`. 
3. We need to match the "flattened"/`::slotted` tree to support this.

There are 2 separate issues which I think should be considered orthogonal to `:has-slotted`:
1. matching out of scope distributed fallback content: there's currently no way to do this with `::slotted` either so this shouldn't be uniquely handled by `has-slotted`.
4. matching against whitespace only: there's also an issue with this using `:empty` so similarly we shouldn't uniquely handle it here.

**Comment by keithamus** [comment]
A quick test tells me the current functionality is correct; slotting a `<slot></slot>` element does not render fallback content, meaning `flattened: false` will observe this, while `flattened: true` does not.

Test case:

`data:text/html,<div><template shadowrootmode="open"><div><template shadowrootmode="open"><slot>Nested Fallback</slot></template><slot></slot></div></template></div>`

**Comment by sorvell** [comment]
> RESOLVED: `:has-slotted` should match when the fallback content is not being displayed

Unless I'm mistaken, `:has-slotted` enables this, which is great:

```html
<padded-with-icon>
  not padded
</padded-with-icon>
<padded-with-icon>
  padded <i-con slot="icon"></i-con>
</padded-with-icon>
```

But this has-a composition won't work because the icon slot inside `padded-with-icon` will not be showing fallback content. This seems pretty unfortunate.

```html
<has-a-padded-with-icon>
  <template shadowrootmode="open">
    <padded-with-icon>
      <slot></slot>
      <slot name="icon" slot="icon"></slot>
    </padded-with-icon>
  </template>
  oops, this is padded
<has-a-padded-with-icon>
```

**Comment by sorvell** [comment]
> RESOLVED: :has-slotted should match when the fallback content is not being displayed

This resolution does not fit with the general [consensus](https://github.com/w3c/csswg-drafts/issues/6867#issuecomment-2368965064) in this issue that `:has-slotted` should behave generally consistent with `::slotted()`, modulo text nodes.

The need here is to know if renderable content has been supplied to the slot. Unfortunately, this is not the same as if it will show fallback content. This is because a `<slot>` *does not* render but *does* prevent fallback content.

The ability to select if fallback content is showing, if specific elements have been slotted, or if slots themselves have been slotted should be handled as follow-on cases since they are (1) less needed and more complex, and (2) could be covered by a functional form and/or a combinator if/when that is possible.

**Proposal** 

`::slotted()` matches the flattened tree so `:has-slotted` should as well. `:has-slotted` should match when `assignedNodes({flatten: true}).length > 0`. This is the same as when slotted(*)` would match except in the case of the default slot which can assign text nodes.

**Comment by justinfagnani** [comment]
I strongly agree with @sorvell here. I think it'll be confusing to have this divergence from `::slotted()` _and_ the composition use case is important.
